jasmine-standalone-2.3.4 (4)/src/Game.js

```diff
@@ -0,0 +1,156 @@
+function Game() {
+  this.frames = [];
+  this.pairs = [];
+}
+
+ // GENERAL METHODS
+
+Game.prototype.isInProgress = function() {
+  return (this._endGameStrike() || this._endGameSpare() || this._lessThan20());
+}
+
+Game.prototype._lastFrame = function() {
+  return this.frames[this.frames.length - 1];
+}
+
+Game.prototype._secondLastFrame = function() {
+  return this.frames[this.frames.length - 2];
+}
+
+Game.prototype._numberOfPairs = function() {
+  return this.pairs.length;
+}
+
+Game.prototype._valueOfPair = function(i) {
+  return Object.values(this.pairs[i])[0];
+}
+
+// ADD FRAME
+
+Game.prototype._addFrame = function() {
+  let framesCount = this.frames.length;
+  let lastFrame = this._lastFrame();
+  let secondLastFrame = this._secondLastFrame();
+
+  if (framesCount === 0 || framesCount === 1) {
```

Comment:
A multi-pronged if statement, and a nested if statement in the last fork! Whenever we see nested conditionals, or conditionals with more than 2 distinct paths, we should be thinking about whether we should refactor our code.

In this case for example, I might follow the process:

Q: What is the outcome of this method?
A: We push a number to the frames collection.

Q: What does the number represent?
A: The frame that the last roll relates to

Q: Why is this complicated?
A: Because we have multiple cases

Q: What are the cases?
A: We only iterate the frame when there have been two rolls, so a roll can be either the first or the second, and also there is a special case relating to the last frame. 

Q: So we have a two to one relationship rolls to frames... sounds like we could abstract that?
A: Hmm. Yes. If frames were a class, and we made it responsible for knowing if this was the first or second roll, we could lose one of the if statements.

Q: Great. And the last frame case... why is that special?
A: Because we can multiple additional rolls if we score a strike or a spare.

Q: Are those additional rolls treated the same as normal rolls? 
A: No. They are only applied as bonuses to the last frame. 

Q: Do we have bonus rolls anywhere else?
A: Yes. All frames can have bonus rolls.

Q: So it's not a special case.
A: But bonus rolls for the last frame don't count towards other frames, because there are no more frames.

Q: Maybe the problem is that we are treating bonuses and rolls as the same thing. 
A: Ah, yes, if our frame class had 'rolls' and 'bonuses' as separate properties, then I'm pretty sure I can treat all frames the same. Then the other conditional goes away. Awesome!

This is obviously an imaginary conversation. But the key things to highlight are:
- When you see nested conditionals or conditionals with multiple pathways, recognise this as a sign that something could be better.
- Refactor it early. The longer you leave it to refactor the harder it will be to unpick.
- Interrogate your code. You could try using the 'Five Whys'. Basically, the goal is to keep challenging your assumptions until you're able to unpick them to something cleaner.